app/jobs/scheduled/ignored_users_summary.rb

```diff
@@ -0,0 +1,45 @@
+module Jobs
+  class IgnoredUsersSummary < Jobs::Scheduled
+    every 1.day
+
+    def execute(args)
+      return unless SiteSetting.ignore_user_enabled
+
+      params = {
+        threshold: SiteSetting.ignored_users_count_message_threshold,
+        gap_days: SiteSetting.ignored_users_message_gap_days,
+      }
+      user_ids = DB.query_single(<<~SQL, params)
+        SELECT ig.ignored_user_id AS user_id
+        FROM ignored_users AS ig
+        WHERE NOT EXISTS (SELECT 1
```

Comment:
I just noticed something quite odd here. There's no `post_id` restriction in your subquery.

It seems like you are using a custom field on the `Post` to avoid notifying moderators twice about the same user? If anything, wouldn't you want to add the custom field on the user rather than the post?

I am also not a huge fan of using Custom Fields in the core application to track data, as you can see the query is a little clumsy. Is there a reason you didn't add a field to `IgnoredUser` such as `summarized_at`?

Reply:
Thanks Robin for the feedback 🙌 

> I just noticed something quite odd here. There's no post_id restriction in your subquery.

Since we are depending on the `post_custom_fields` table to ensure idempotency for system message creation, and the idempotency key here being: `'summary_sent_for_ignored_user' + 'user_id' + 'created_at'`, can you elaborate more on why do we need `post_id`? Or rather will there be a data integrity issue if we didn't bind it by `post_id`? 

> It seems like you are using a custom field on the Post to avoid notifying moderators twice about the same user? 

Yes 👍 

> If anything, wouldn't you want to add the custom field on the user rather than the post?

I think that's cleaner, or rather the relationship is more straight forward. I have no strong opinions of going with one of the two approaches, I think it yields the same result for us, one more rational than the other.

> I am also not a huge fan of using Custom Fields in the core application to track data

Yes, I didn't feel super good about it too, we had few choices being:

1. Find a way to track this data by maybe introducing a generic table for tracking idempotent operations, use that table for this and everything else of the same nature in the future.
2. Build upon and use an already existing persistence model (aka: `PostCustomField` or `UserCustomField` tables).
3. Use volatile storage medium (Redis) but I feel that's unreliable.

Perhaps you have other options?

> Is there a reason you didn't add a field to IgnoredUser such as summarized_at?

The relationship is as: `User --> has many --> Ignored Users`, if I add a `summarised_at` on the `IgnoredUser: X`, for example, I need to add that on all `IgnoredUser: X` records, right? Or how would that work?

I think we chose the `PostCustomField` because we know that one Post (System message) is going to ever have one unique combination of:  `'summary_sent_for_ignored_user' + 'user_id' + 'created_at'`.

Reply:
To summarise the above and move this forward,

1. Should we go with the `UserCustomField`?
2. Any other possible or cleaner options than ensuring idempotency through `*CustomField` tables?

Reply:
> can you elaborate more on why do we need post_id

To be clear, you don't for your code to work. But it indicates to me that the fields are not `Post` specific, they are user specific. That's why I brought it up :)

(BTW while speaking of idempotent operations, custom fields is not great. It has known issues where you can insert the same value twice concurrently and Discourse treats it as an array. We are going to fix this.) 

> The relationship is as: User --> has many --> Ignored Users, if I add a summarised_at on the IgnoredUser: X, for example, I need to add that on all IgnoredUser: X records, right? Or how would that work?

Yes, let's say you added a column `summarized_at DATETIME` to the `ignored_users` table. When a summary is sent for that user you can update it atomically:

`IgnoredUser.where(ignored_user_id: user_id).update_all(summarized_at: Time.zone.now)`

And your query is a lot simpler now

```sql
SELECT ig.ignored_user_id AS user_id
FROM ignored_users AS ig
WHERE (ig.created_at + ':gap_days DAYS'::INTERVAL) > CURRENT_TIMESTAMP
GROUP BY ig.ignored_user_id
HAVING COUNT(*) > :threshold
```

The minor downside of this design is the extra column, but it's significantly less overhead than using the user custom fields table.

Reply:
I think the query should be 

``` sql
SELECT ignored_user_id AS user_id
FROM ignored_users
WHERE (summarized_at + ':gap_days DAYS'::INTERVAL) > CURRENT_TIMESTAMP
GROUP BY ignored_user_id
HAVING COUNT(*) > :threshold
```

I changed the `where` condition from using `created_at` to using `summarized_at` since we only want the user_ids of users that haven't already been "summarized" in the past X days.

Reply:
Yes oops, I meant `summarized_at` (wrote this from memory.)

Reply:
I love this solution ❤️ , though this will fail if:

1. A User has been ignored by 5 other users (passing threshold)
2. We send a System message, and do the `update_all`
3. The same User has been un-ignored from the same 5 other users before the `gap_days` end
4. With every un-ignore, we delete the matching `ignored_users` db record.
5. Now, if the user to be ignored _yet again_ by 5 other users, we will send a System message which we aren't supposed to sent (because we have no records in the `ignored_user` table to give us a hint).

I wouldn't say it is a big deal because of the rare possibility 😃 but it is a scenario that can happen.

Reply:
We could soft delete the records instead of hard deleting them ;) That way we'd keep the `summarized_at`. Another solution would be to have a new table for summarized users.

Reply:
I thought of soft delete, but didn't want to mention it 😅

Reply:
Nice catch, but I agree it seems so unlikely. Let's go with this, and if in the future we get any reports of this will split it off to be a new table.

Reply:
Let's do it 🙌